src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,349 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal interface ITypeImportCompletionService : IWorkspaceService
+    {
+        void GetAccessibleTopLevelTypesFromPEReference(
+            Solution solution,
+            Compilation compilation,
+            PortableExecutableReference peReference,
+            Action<CompletionItem> handleAccessibleItem,
+            CancellationToken cancellationToken);
+
+        Task GetAccessibleTopLevelTypesFromCompilationReferenceAsync(
+            Solution solution,
+            Compilation compilation,
+            CompilationReference compilationReference,
+            Action<CompletionItem> handleAccessibleItem,
+            CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Get all the top level types from given project. This method is intended to be used for 
+        /// getting types from source only, so the project must support compilation. 
+        /// For getting types from PE, use <see cref="GetAccessibleTopLevelTypesFromPEReference"/>.
+        /// </summary>
+        Task GetAccessibleTopLevelTypesFromProjectAsync(
+            Project project,
+            Action<CompletionItem> handleAccessibleItem,
+            CancellationToken cancellationToken);
+    }
+
+    [ExportWorkspaceServiceFactory(typeof(ITypeImportCompletionService), ServiceLayer.Editor), Shared]
+    internal sealed class TypeImportCompletionService : IWorkspaceServiceFactory
+    {
+        private readonly ConcurrentDictionary<string, ReferenceCacheEntry> _peItemsCache
```

Comment:
so this cache lives whole VS session. I think you might need to run VAL build RPS just to make sure that memory is okay with RPS.

Reply:
I'm curious about the workspace cache service. When will it call cache flush?

Reply:
It flushes when VS is "low on memory". It's more a "we're about to sink, everybody start bailing!" event than something intended as a regular cache expiration mechanism.